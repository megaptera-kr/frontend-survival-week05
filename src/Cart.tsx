import { useLocalStorage } from 'usehooks-ts';

import useCreateOrder from './hooks/useCreateOrder';

import Food from './types/Food';
import Receipt from './types/Receipt';
import CartItem from './CartItem';
import OrderBtn from './OrderBtn';

type CartProps = {
  setReceipt: (receipt: Receipt) => void;
};

export default function Cart({ setReceipt }: CartProps) {
  const [selectedFoods, setFoods] = useLocalStorage<Food[]>('cart', []);

  const totalPrice = selectedFoods.reduce((acc, food) => food.price + acc, 0);

  const createOrder = useCreateOrder();

  const handleClickCancel = (index: number) => {
    const food = selectedFoods.filter((_, i) => i !== index);
    setFoods(food);
  };

  const handleOrder = async () => {
    if (!selectedFoods.length) {
      return;
    }
    const receipt = await createOrder(selectedFoods);
    setReceipt(receipt);
    setFoods([]);
  };

  return (
    <div>
      <h2>점심 바구니</h2>
      <ul style={{ width: '20%' }}>
        {selectedFoods?.map((food, index) => {
          const key = `${food}-${index}`;
          return (
            <CartItem
              key={key}
              food={food}
              index={index}
              handleClickCancel={handleClickCancel}
            />
          );
        })}
      </ul>
      <OrderBtn totalPrice={totalPrice} handleOrder={handleOrder} />
    </div>
  );
}
