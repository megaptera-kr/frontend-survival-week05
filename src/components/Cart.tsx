import { useLocalStorage } from 'usehooks-ts';
import Food from '../types/Food';
import MenuItems from './MenuItems';
import Receipt from '../types/Receipt';
import useOrder from '../hooks/useOrder';

type CartProps = {
  setReceipt: (value: Receipt) => void;
}

export default function Cart({ setReceipt }: CartProps) {
  const [selectedFoods, setFoods] = useLocalStorage<Food[]>('cart', []);

  const totalPrice = selectedFoods.reduce((acc, food) => (acc + food.price), 0);

  const handleClickOrder = () => {
    const receipt = useOrder();
    setReceipt(receipt);
    setFoods([]);
  };

  return (
    <>
      <h2>점심 바구니</h2>
      <div>
        <MenuItems menu={selectedFoods} buttonType="삭제" />
      </div>
      <button
        type="button"
        onClick={handleClickOrder}
      >
        합계:
        {' '}
        {totalPrice.toLocaleString()}
        원 주문
      </button>
    </>
  );
}
