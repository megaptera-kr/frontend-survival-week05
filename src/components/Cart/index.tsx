import { useContext } from 'react';
import { postOrders } from '../../apis';
import { OrderContext } from '../../contexts/OrderContext';
import { Menu } from '../../types/restaurant';
import toLocaleCurrency from '../../utils/format';

export default function Cart() {
  const { cart, setCart, setReceipt } = useContext(OrderContext);
  const totalPrice = cart.reduce((prev, cur) => prev + cur.price, 0);

  const handleClickRemoveMenu = (menu: Menu) => () => {
    setCart?.(cart.filter((cartMenu) => cartMenu.id !== menu.id));
  };

  const handleClickOrderMenu = async () => {
    const res = await postOrders({
      menu: cart,
      totalPrice,
    });

    setReceipt?.(res);
    setCart?.([]);

    setTimeout(() => { setReceipt?.(null); }, 8000);
  };

  return (
    <div>
      <h2>점심 바구니</h2>
      <ul>
        {cart?.map((menu) => (
          <li key={menu.id}>
            {menu.name}
            (
            {toLocaleCurrency(menu.price)}
            원)
            <button type="button" onClick={handleClickRemoveMenu(menu)}>취소</button>
          </li>
        ))}
      </ul>
      <button type="button" onClick={handleClickOrderMenu}>
        합계:
        {' '}
        {toLocaleCurrency(totalPrice)}
        원 주문
      </button>
    </div>
  );
}
