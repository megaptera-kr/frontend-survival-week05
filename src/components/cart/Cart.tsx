import useCart from '../../hooks/useCart';
import CartTable from './cartTable/CartTable';
import CartUtils from './cartUtils/Cartutils';
import useCreateOrder from './hooks/useCreateOrder';
import { calcTotalPrice } from '../../utils/cart';
import { Receipts } from '../../types/receipt';

interface CartProps {
  createReceipts: (receipts: Receipts) => void;
}

export default function Cart({ createReceipts }: CartProps) {
  const { cartItems, handleRemoveCart, handleClearCart } = useCart();
  const { createOrder } = useCreateOrder();

  const handleCreateOrder = async () => {
    const receipt = await createOrder(cartItems);
    createReceipts(receipt);
    handleClearCart();
  };

  const totalPrice = calcTotalPrice(cartItems);
  return (
    <article>
      <h2>장바구니</h2>
      <CartUtils
        totalPrice={totalPrice}
        handleClearCart={handleClearCart}
        handleOrder={handleCreateOrder}
      />
      <CartTable
        cartItems={cartItems}
        handleRemoveCart={handleRemoveCart}
      />
    </article>
  );
}
