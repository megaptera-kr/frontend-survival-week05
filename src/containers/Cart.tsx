import MenuItem from '../components/MenuItem';
import useFetchOrders from '../hooks/useFetchOrders';
import { useCartStorage } from '../hooks/useStorage';
import createReceiptData from '../utils/createReceiptData';

export const testFn = () => {};

function Cart() {
  const { cart, handleRemoveCart } = useCartStorage();

  const { payment } = useFetchOrders();

  const receiptData = createReceiptData(cart);

  return (
    <article data-testid="Cart">
      <h2>점심 바구니</h2>
      <ul>
        {cart.map((menu, index) => (
          <li data-testid="CartItem" key={`${menu.id}_${index}`}>
            <MenuItem
              menuItem={menu}
              index={index}
              btnLabel="취소"
              onClick={handleRemoveCart}
            />
          </li>
        ))}
      </ul>
      <button data-testid="Payment" type="button" onClick={payment} disabled={!cart.length}>
        {`합계: ${receiptData.totalPrice.toLocaleString()}원 주문`}
      </button>
    </article>
  );
}

export default Cart;
