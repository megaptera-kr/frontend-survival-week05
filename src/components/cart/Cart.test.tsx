import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Cart from './Cart';
import '../../tests/stub/localStorage';
import { foods } from '../../../fixtures';

const context = describe;

describe('Cart', () => {
  const createReceipts = jest.fn();

  function renderCart() {
    return render(<Cart createReceipts={createReceipts} />);
  }

  function setLocalStorage <T>(id: string, data: T) {
    window.localStorage.setItem(id, JSON.stringify(data));
  }

  beforeEach(() => {
    setLocalStorage('kioskCart', foods);
  });

  it('render correctly', async () => {
    const { asFragment } = renderCart();

    await screen.findByText('짜장면');

    expect(asFragment()).toMatchSnapshot();
  });

  context('when click clear button', () => {
    it('cart is empty', async () => {
      renderCart();

      const clearBtn = screen.getByRole('button', { name: '비우기' });
      userEvent.click(clearBtn);

      await waitFor(() => expect(screen.queryByText('짜장면')).toBeNull());
      await waitFor(() => expect(screen.queryByText('짬뽕')).toBeNull());
      await waitFor(() => expect(screen.queryAllByRole('cell').length).toBe(0));
    });
  });

  context('when click remove button', () => {
    it('food is will be removed from the cart', async () => {
      renderCart();

      const removeBtn = await screen.findByRole('button', { name: /짜장면제거/ });
      userEvent.click(removeBtn);

      await waitFor(() => expect(screen.queryByText('짜장면')).toBeNull());
      await waitFor(() => expect(screen.queryAllByRole('cell').length).toBe(3));
    });
  });

  context('when click order button', () => {
    it('cart is emepty', async () => {
      renderCart();

      const orderBtn = await screen.findByRole('button', { name: '주문하기' });
      userEvent.click(orderBtn);

      await waitFor(() => expect(screen.queryByText('짜장면')).toBeNull());
      await waitFor(() => expect(screen.queryByText('짬뽕')).toBeNull());
      await waitFor(() => expect(screen.queryAllByRole('cell').length).toBe(0));
    });
  });
});
