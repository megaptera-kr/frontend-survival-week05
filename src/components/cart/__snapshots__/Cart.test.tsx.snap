// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart render correctly 1`] = `
<DocumentFragment>
  <article>
    <h2>
      장바구니
    </h2>
    <div>
      <button
        type="button"
      >
        비우기
      </button>
      <button
        aria-label="주문하기"
        type="button"
      >
        합계: 13000원 주문
      </button>
    </div>
    <table
      style="border: 1px solid black; background-color: black;"
    >
      <caption>
        점심 바구니
      </caption>
      <thead
        style="background-color: white;"
      >
        <tr>
          <th
            scope="col"
          >
            음식 명
          </th>
          <th
            scope="col"
          >
            음식 가격
          </th>
          <th
            scope="col"
          >
            제거하기
          </th>
        </tr>
      </thead>
      <tbody
        style="background-color: white;"
      >
        <tr>
          <td>
            짜장면
          </td>
          <td>
            8000
          </td>
          <td>
            <button
              type="button"
            >
              짜장면제거
            </button>
          </td>
        </tr>
        <tr>
          <td>
            짬뽕
          </td>
          <td>
            5000
          </td>
          <td>
            <button
              type="button"
            >
              짬뽕제거
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </article>
</DocumentFragment>
`;
